transactron/lib/fifo.py

```diff
@@ -114,6 +118,176 @@ def _() -> None:
+
+    Attributes
+    ----------
+    read: Method
+        Reads from the FIFO. Accepts a single argument `count`, denoting the number of elements
+        to be read. Returns `count` elements, or less, if the queue has less than `count` elements.
+        Ready only if the FIFO is not empty.
+    peek: Method
+        Returns the elements at the front (without removing them). Ready only if the FIFO
+        is not empty. The method is nonexclusive.
+    write: Method
+        Writes to the FIFO. Ready only if the FIFO has enough free slots to accept the write.
+    clear: Method
+        Clears the FIFO entries. Has priority over `read` and `write` methods.
+    """
+
+    def __init__(
+        self, shape: ShapeLike, depth: int, read_width: int, write_width: Optional[int], *, src_loc: int | SrcLoc = 0
+    ) -> None:
+        """
+        Parameters
+        ----------
+        shape: ShapeLike
+            Shape of the data stored in the queue.
+        depth: int
+            Depth of the FIFO. The number of elements which can be stored in the queue
+            is `depth * max(read_width, write_width)`.
+        read_width: int
+            Number of elements which can be simultaneously read from the queue.
+        write_width: int, optional
+            Number of elements which can be simultaneously written to the queue.
+            If omitted, it is assumed to be equal to `read_width`.
+        src_loc: int | SrcLoc
+            How many stack frames deep the source location is taken from.
+            Alternatively, the source location to use instead of the default.
+        """
+        if write_width is None:
+            write_width = read_width
+
+        self.shape = shape
+        self.read_layout = data.StructLayout(
+            {"count": range(read_width + 1), "data": data.ArrayLayout(shape, read_width)}
+        )
+        self.write_layout = data.StructLayout(
+            {"count": range(write_width + 1), "data": data.ArrayLayout(shape, write_width)}
+        )
+        self.read_width = read_width
+        self.write_width = write_width
+        self.depth = depth
+
+        self.read = Method(i=[("count", range(read_width + 1))], o=self.read_layout, src_loc=src_loc)
+        self.peek = Method(o=self.read_layout, nonexclusive=True, src_loc=src_loc)
+        self.write = Method(i=self.write_layout, src_loc=src_loc)
+        self.clear = Method(src_loc=src_loc)
+
+    def elaborate(self, platform):
+        m = TModule()
+
+        max_width = max(self.read_width, self.write_width)
+
+        storage = [memory.Memory(shape=self.shape, depth=self.depth, init=[]) for _ in range(max_width)]
+
+        for i, mem in enumerate(storage):
+            m.submodules[f"storage{i}"] = mem
+
+        write_ports = [mem.write_port() for mem in storage]
+        read_ports = [mem.read_port(domain="sync", transparent_for=[port]) for mem, port in zip(storage, write_ports)]
```

Comment:
Is the transparency needed?

Reply:
I believe so, same as in `BasicFifo`. Explanation: if the FIFO is empty and a `write` is performed in cycle 1, then if `read` is called in cycle 2, it needs to return the value written in cycle 1, which is provided thanks to transparency. If you remove transparency, the test fails.